src/IO/S3/copyDataToS3.cpp

```diff
@@ -0,0 +1,682 @@
+                        log,
+                        "Single part upload failed with error {} for Bucket: {}, Key: {}, Object size: {}, will retry with multipart upload",
+                        outcome.GetError().GetExceptionName(),
+                        dest_bucket,
+                        dest_key,
+                        size);
+                    performMultipartUpload();
+                    break;
+                }
+                else if (outcome.GetError().GetErrorType() == Aws::S3::S3Errors::NO_SUCH_KEY)
+                {
+                    /// For unknown reason, at least MinIO can respond with NO_SUCH_KEY for put requests
+                    LOG_INFO(
+                        log,
+                        "Single part upload failed with NO_SUCH_KEY error for Bucket: {}, Key: {}, Object size: {}, will retry",
+                        dest_bucket,
+                        dest_key,
+                        request.GetContentLength());
+                    /// will retry
+                }
+                else
+                {
+                    throw S3Exception(
+                        outcome.GetError().GetErrorType(),
+                        "Message: {}, Key: {}, Bucket: {}, Object size: {}",
+                        outcome.GetError().GetMessage(),
+                        dest_key,
+                        dest_bucket,
+                        request.GetContentLength());
+                }
+            }
+        }
+
+        void performMultipartUpload() { UploadHelper::performMultipartUpload(offset, size); }
+
+        std::unique_ptr<Aws::AmazonWebServiceRequest> fillUploadPartRequest(size_t part_number, size_t part_offset, size_t part_size) override
+        {
+            auto read_buffer = std::make_unique<LimitSeekableReadBuffer>(create_read_buffer(), part_offset, part_size);
+
+            /// Setup request.
+            auto request = std::make_unique<Aws::S3::Model::UploadPartRequest>();
+            request->SetBucket(dest_bucket);
+            request->SetKey(dest_key);
+            request->SetPartNumber(static_cast<int>(part_number));
+            request->SetUploadId(multipart_upload_id);
+            request->SetContentLength(part_size);
+            request->SetBody(std::make_unique<StdStreamFromReadBuffer>(std::move(read_buffer)));
+
+            /// If we don't do it, AWS SDK can mistakenly set it to application/xml, see https://github.com/aws/aws-sdk-cpp/issues/1840
+            request->SetContentType("binary/octet-stream");
+
+            return request;
+        }
+
+        String processUploadPartRequest(Aws::AmazonWebServiceRequest & request) override
+        {
+            auto & req = typeid_cast<Aws::S3::Model::UploadPartRequest &>(request);
+
+            ProfileEvents::increment(ProfileEvents::S3UploadPart);
+
+            auto outcome = client_ptr->UploadPart(req);
+            if (!outcome.IsSuccess())
+            {
+                abortMultipartUpload();
```

Comment:
> However, if any part uploads are currently in progress, those part uploads might or might not succeed. As a result, it might be necessary to abort a given multipart upload multiple times in order to completely free all storage consumed by all parts.

As I see, this call is concurrent with other UploadPart calls. In that case it might be useless.

Reply:
I think I have to check the result of `aboutMultipartUpload` - and if it didn't succeed then I need to call `aboutMultipartUpload` again in the end. I'll change the code.

Reply:
done